### PR TITLE
[CMake/Bazel] Support usage of opt driver as a library

### PR DESCRIPTION
**Pull request by wsmoses**
In Bazel, Clang current separates the clang executable into a clang-driver library, and the actual clang executable. This allows downstream users to make their own variations of clang, without having to redo/maintain separate build pipelines.

This adds the same for opt for both CMake and Bazel.

### CODE REVIEW

**Review by efriedma-quic — CHANGES_REQUESTED** [review]
If we're going to make opt-driver a library, it should be a library for both CMake and Bazel builds.  The set of available libraries shouldn't depend on how you build LLVM.

**Comment by wsmoses** [comment]
I'm fine with that but FYI I emulated the clang approach, which does have this in Bazel (https://github.com/llvm/llvm-project/blob/03a61d34ebf4f8eeaa6861bec3ab39c75bb41778/utils/bazel/llvm-project-overlay/clang/BUILD.bazel#L2275) but not CMake (https://github.com/llvm/llvm-project/blob/03a61d34ebf4f8eeaa6861bec3ab39c75bb41778/clang/tools/driver/CMakeLists.txt#L26)

**Comment by wsmoses** [comment]
@efriedma-quic added a CMake Library for opt. I presume we should add one for clang as well, but that should probably be a separate PR.

**Comment by efriedma-quic** [comment]
Looks okay to me, but I'm a bit worried there's something I'm missing, so I'd like a second reviewer to chime in.

**Comment by MaskRay** [comment]
> This allows downstream users to make their own variations of clang, without having to redo/maintain separate build pipelines.

Consider naming the project, so if other people do any change, they can find the project name from the commit message...

It may be worth testing the following configurations:

```
default
-DBUILD_SHARED_LIBS=on
-DLLVM_LINK_LLVM_DYLIB=on
```

**Review by MaskRay — APPROVED** [review]

File: llvm/tools/opt/CMakeLists.txt [hunk 1]
```diff
@@ -29,12 +29,24 @@ set(LLVM_LINK_COMPONENTS
   Passes
   )
 
-add_llvm_tool(opt
+# We don't want to link this into libLLVM
+add_llvm_library(LLVMOptDriver
+  STATIC
```
Comment (nico) [inline llvm/tools/opt/CMakeLists.txt:34]: The usual llvlm style is to have one target per directory. If this is now a lib, could you move this into llvm/tools/opt/lib? Then you don't need the  PARTIAL_SOURCES_INTENDED below.

**Comment by nico** [comment]
Also, this doesn't build, see e.g. here: https://lab.llvm.org/buildbot/#/builders/217/builds/35051

**Comment by nico** [comment]
Reverted in be08be5d5de97cd593fb99affa1fa994d104eb70 for now.

**Comment by nico** [comment]
Whoops, sorry, the build failure was due to something else! I'll reland, apologies. Please do move the lib to its own dir though.

**Comment by nico** [comment]
Relanded in 31359840247b7d458ea8104373eea3f0cef95e16. Apologies again.

**Comment by wsmoses** [comment]
Will do!

**Pull request merged**